app/timecode/views.py

```diff
@@ -0,0 +1,216 @@
+
+    def calculate_resp(self, from_subs, from_doc):
+        d = difflib.SequenceMatcher(lambda x: x in " ")
+        d.set_seqs(from_subs, from_doc)
+        diff = d.get_opcodes()
+        resp = {}
+        i = 0
+        l = len(diff)
+        while i < l:
+            _last = i == (l - 1)
+            (tag, i1, i2, j1, j2) = diff[i]
+            try:
+                if _last:
+                    t_start = self.time_by_word_sub[i1 - 1]
+                    t_end = self.duration
+                else:
+                    t_start, t_end = self.find_start_end_by_idxs(i1, i2)
+            except Exception as e:
+                current_app.logger.error(e)
+            if tag == 'equal' or tag == 'replace':
+                for (pos, t) in approximate_doc_time(j1, j2, t_start, t_end):
+                    resp[pos] = RespItem(t, pos)
+            # for insert we can pake time or from prev delete or from next delete if have
+            elif tag == 'insert':
+                try:
+                    p_tag, p_i1, p_i2, p_j1, p_j2 = diff[i - 1]
+                    n_tag, n_i1, n_i2, n_j1, n_j2 = diff[i + 1]
+                    if p_tag == 'delete':
+                        t_start = self.time_by_word_sub[p_i1]
+                        t_end = self.time_by_word_sub[i2]
+                    elif n_tag == 'delete':
+                        if i + 2 < l and diff[i + 2][0] == "insert":
+                            i += 1
+                            continue
+                        t_start = self.time_by_word_sub[i1]
+                        t_end = self.time_by_word_sub[n_i2]
+                    elif i1 == i2:
+                        t_start = self.time_by_word_sub[p_i2]
+                        t_end = self.time_by_word_sub[i2]
+                    else:
+                        t_start = self.time_by_word_sub[i1]
+                        t_end = self.time_by_word_sub[i2]
+
+                    for (pos, t) in approximate_doc_time(j1, j2, t_start, t_end):
+                        resp[pos] = RespItem(t, pos)
+                except Exception as e:
+                    current_app.logger.error('Calculate response for iteration num %s. Exception %s' % (i, e))
+            i += 1
+        return resp
+
+
+class RespItem(dict):
+    def __init__(self, t, idx):
+        dict.__init__(self, timeCode=t, index=idx)
+
+
+def approximate_doc_time(j1, j2, s_time, e_time):
+    try:
+        w_len = j2 - j1
+        coef = 0 if w_len == 0 else (e_time - s_time) / w_len
+
+        resp = []
+        t_offset = s_time
+        w_count = 0
+        while w_count < w_len:
+            resp.append((j1 + w_count, round((t_offset + w_count * coef) / 1000, 1)))
+            w_count += 1
+        return resp
+    except Exception as e:
+        current_app.logger.error(e)
+
+
+# must be synchronized with client side clean text cause we use letter position for find timestamp
+def clear_str(text):
+    pattern = re.compile(r'[".,\/#!$%\^&\*;:{}=\-_`~()\[\]‘’”“]')
```

Comment:
You can compile this once and not each time you call this function

Reply:
fix